src/dstack/_internal/utils/remote.py

```diff
@@ -0,0 +1,164 @@
+import io
+import json
+import time
+from contextlib import contextmanager
+from typing import Any, Dict, Generator, List
+
+import paramiko
+
+from dstack._internal.core.errors import ProvisioningError
+from dstack._internal.core.models.instances import (
+    Disk,
+    Gpu,
+    InstanceType,
+    Resources,
+)
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+SSH_CONNECT_TIMEOUT = 10
+
+DSTACK_SHIM_ENV_FILE = "dstack-shim.env"
+
+
+def sftp_upload(client: paramiko.SSHClient, path: str, body: str) -> None:
+    try:
+        sftp = client.open_sftp()
+        channel = sftp.get_channel()
+        if channel is not None:
+            channel.settimeout(10)
+        sftp.putfo(io.BytesIO(body.encode()), path)
+        sftp.close()
+    except (paramiko.SSHException, OSError) as e:
+        raise ProvisioningError(e)
+
+
+def upload_envs(client: paramiko.SSHClient, working_dir: str, envs: Dict[str, str]) -> None:
+    envs["DSTACK_SERVICE_MODE"] = "1"  # make host_info.json on start
+    dot_env = "\n".join(f'{key.upper()}="{value.strip()}"' for key, value in envs.items())
+    tmp_file_path = f"/tmp/{DSTACK_SHIM_ENV_FILE}"
+    sftp_upload(client, tmp_file_path, dot_env)
+    try:
+        _, stdout, _ = client.exec_command(
+            f"sudo mv {tmp_file_path} {working_dir}/{DSTACK_SHIM_ENV_FILE}", timeout=10
+        )
+        output = stdout.read()
+        if output:
+            logger.warning("Cannot move %s: %s", tmp_file_path, output.decode())
+    except (paramiko.SSHException, OSError) as e:
+        raise ProvisioningError(e)
```

Comment:
Don't wrap catched exceptions with custom exceptions to avoid hiding original `__cause__`, `__context__`, and `__traceback__`. To raise custom exceptions with retaining original context use `raise ProvisioningError() from e`

Reply:
Fixed!
I want to keep the `paramiko` and `socket` exceptions hidden from `process_instances`. So, I wrapped exceptions in ProvisioningError.